File: 0733-flood-fill/0733-flood-fill.py
class Solution(object):
    def floodFill(self, image, sr, sc, color):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type color: int
        :rtype: List[List[int]]
        """
        m, n = len(image), len(image[0])
        old_color = image[sr][sc]
        
        if old_color == color:
            return image
        def dfs(r, c):
            if image[r][c] == old_color:
                image[r][c] = color
                if r >= 1: dfs( r - 1, c)
                if r + 1 < m: dfs(r + 1, c)
                if c >= 1 : dfs (r, c - 1)
                if c + 1 < n: dfs(r, c + 1)
                    
        dfs(sr, sc)
        return image

File: 0733-flood-fill/0733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    let visited = new Set()
    const oldColor = image[sr][sc]
    
    fillColor(image, sr,sc, color, oldColor, visited)
    
    return image
};

const inbound = (r, c, er, ec) => {
    return r >= 0 && c >= 0 && er > r && ec > c
}

const fillColor = (image, sr, sc, color, old, visited) => {
    const pos = sr + ',' + sc
    if (!inbound(sr, sc, image.length, image[0].length)) return 
    if (pos in visited) return 
    visited.add(pos)
    if (image[sr][sc] !== old || image[sr][sc] === color) return
    image[sr][sc] = color
    fillColor(image, sr - 1, sc, color, old, visited)
    fillColor(image, sr + 1, sc, color, old, visited)
    fillColor(image, sr, sc - 1, color, old, visited)
    fillColor(image, sr, sc + 1, color, old, visited)
}